core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,9 +1158,14 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
```

Comment:
@pengwei-li, can you add some logging to the `ZKCheckedEphemeral` code to help us understand what's happening. Maybe:

```scala
              case Code.OK =>
                if (stat.getEphemeralOwner != zkHandle.getSessionId) {
                  debug(s"Failed to create ephemeral node as session ids differ, ${stat.getEphemeralOwner} vs ${zkHandle.getSessionId}")
                  setResult(Code.NODEEXISTS)
                }
                else {
                  debug(s"Node created by the same session already exists, nothing to do")
                  setResult(Code.OK)
                }
```